Blocked: ECDSA Signature Verification in PKC

suyash67 opened the pull request:
# Description

This PR adds ECDSA signature verification in the private kernel circuit. This is necessary as the final transaction from a client (i.e. the user) needs validation that the intended user has initiated the intended transaction with correct inputs. 

#### C++ Changes Overview

To implement ECDSA signature verification in the private kernel circuit, we add `signing_key` to the `signed_tx_request`:
```c++
template <typename NCT> struct SignedTxRequest {
    TxRequest<NCT> tx_request{};
    secp256k1_point signing_key;
    Signature signature{};
}
```
From here on, we will use $S\in\mathbb{G}_{\textsf{secp256k1}}$ to denote the signing key. We use the ECDSA signing algorithm to ensure that the existing Ethereum public keys can sign transactions without needing to generate any new keys. The private key $s\in\mathbb{F}\_{q, \textsf{secp256k1}}$ corresponding to the signing key $S$ is used to sign a transaction.
```ts
message = tx_request.hash();
signature = crypto::ecdsa::construct_signature(message, private_key)
```

This signature then needs to be verified in the private kernel circuit. 
```ts
verification_result = crypto::ecdsa::verify_signature(message, signing_key, signature)
```

More on the ECDSA algorithms can be found in [this](https://discourse.aztec.network/t/ecdsa-signature-verification-in-private-kernel-circuit/177) discourse post. 

In addition to the signature verification, we also need to check if the Ethereum address correctly maps to the public key. 
```ts
compute_ethereum_address(signed_tx_request.signing_key) == tx_request.from
```
We assume that the `from` field in the `tx_request` is an ethereum address and ensure in the circuit that this address indeed was computed from `signing_key`.

#### TS Changes Overview

Obviously, we need to add a `signingKey` to `signedTxRequest` to replicate the ABI structure from C++. For this, we implement a `keyStore` for the secp256k1 curve. Further, while running the PKC, it may not always have access to the Ethereum public key of the sender. We might only have access to the Ethereum address and the app sends a signature to our TS code to initiate PKC. So we needed to implement public key recovery from ECDSA signature in C++. We expose a C++ method called `crypto::ecdsa::recover_public_key` to TS for TS to compute the signing key from a signature. 

# Checklist:

- [x] I have reviewed my diff in github, line by line.
- [x] Every change is related to the PR description.
- [x] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this pull request to the issue(s) that it resolves.
- [x] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [x] The branch has been merged or rebased against the head of its merge target.
- [x] I'm happy for the PR to be merged at the reviewer's next convenience.

suyash67 committed:
[SQUASHED] ECDSA in private kernel circuit.

fix signed tx re `to_circuit_type`.

ecdsa init.

typedef fr.

ecdsa wip.

ecdsa progress. test wip.

Fix sig verification.

Fix

Fix `keccak(public_key) == address`.

Use `assert_is_zero`.

Fix other tests and update bb.

Add a comment.

Remove `TX_FROM` and `TX_TO` constants.

remove unused commented line.

Rebase fix.

Move sig stuff to `validate_inputs`.

Disable `USE_TURBO`

Add sig ver in native pkc.

Move `compute_eth_address` to utils.

Mike's suggestions.

update bb.

update bb.

Add errors codes for ecdsa.

Minor fixes and update bb.

suyash67 committed:
Add ecdsa ts.

suyash67 committed:
Get circuits.js working.

Update bb.

Rename `one` to `generator`.

suyash67 committed:
important TODO re crs.

suyash67 committed:
Add secp256k1 prelims.

suyash67 committed:
Add secp256k1 key store.

iAmMichaelConnor commented on circuits/cpp/bootstrap.sh:
```diff
@@ -6,7 +6,7 @@ set -eu
 
 # Update the submodule
-git submodule update --init --recursive
+# git submodule update --init --recursive
```
Is this something that will need to be un-commented, once you're done?

suyash67 replied:
That's right.

iAmMichaelConnor commented on circuits/cpp/src/aztec3/circuits/abis/c_bind.test.cpp:
```diff
@@ -57,6 +57,7 @@ TEST(abi_tests, hash_tx_request)
     // Construct TxRequest with some randomized fields
     TxRequest<NT> tx_request = TxRequest<NT>{
         .from = NT::fr::random_element(),
+        .from_public_key = NT::secp256k1_point::random_element(),
```
As discussed, let's move this to the private kernel circuit's inputs struct

iAmMichaelConnor commented on circuits/cpp/src/aztec3/circuits/abis/tx_request.hpp:
```diff
@@ -14,11 +14,13 @@ using aztec3::utils::types::CircuitTypes;
     typedef typename NCT::boolean boolean;
 
     address from = 0;
+    secp256k1_point from_public_key;
```
As discussed, let's move this to the signed tx request struct

suyash67 replied:
Okay so I did try moving this to `signed_tx_request`. It leads to a lot of changes on the TS side and I think its better to do that as its own PR).

iAmMichaelConnor commented on circuits/cpp/src/aztec3/circuits/kernel/private/native_private_kernel_circuit.cpp:
```diff
@@ -301,6 +303,42 @@ void validate_inputs(DummyComposer& composer, PrivateInputs<NT> const& private_i
 
+    {
+        // Verify public key hash matches ethereum address of the sender
+        // TODO(Suyash): Do we want to perform this check only for base case?
```
Correct - it's only needed for the base case. (For the "initial private kernel circuit", when we create it).
A signed tx request won't even be sent to the 'non initial' version of the private kernel circuit.

suyash67 replied:
right, thanks. I've added a comment to move this to "initial" pkc once its ready.

iAmMichaelConnor commented on circuits/cpp/src/aztec3/circuits/kernel/private/native_private_kernel_circuit.cpp:
```diff
@@ -301,6 +303,42 @@ void validate_inputs(DummyComposer& composer, PrivateInputs<NT> const& private_i
+    {
+        // Verify public key hash matches ethereum address of the sender
+        // TODO(Suyash): Do we want to perform this check only for base case?
+        // TODO(Suyash): I think its okay to check this for each kernel iteration.
```
It's ok for now, but we can remove it from the 'non initial' kernel circuit, once we write that.

iAmMichaelConnor commented on circuits/cpp/src/aztec3/circuits/abis/signed_tx_request.hpp:
```diff
@@ -51,10 +46,40 @@ template <typename NCT> struct SignedTxRequest {
+        return NCT::compress(messages, aztec3::SIGN_TX_REQUEST);
+    }
+
+    void compute_signature(const NativeTypes::secp256k1_fr& private_key)
+    {
+        static_assert((std::is_same<NativeTypes, NCT>::value));
+        auto signing_message = compute_signing_message().to_buffer();
+        std::string signing_message_str(signing_message.begin(), signing_message.end());
+        crypto::ecdsa::key_pair<NativeTypes::secp256k1_fr, NativeTypes::secp256k1_group> account;
+        account.private_key = private_key;
+        account.public_key = tx_request.from_public_key;
+
+        signature = crypto::ecdsa::construct_signature<Sha256Hasher,
+                                                       NativeTypes::secp256k1_group::Fq,
+                                                       NativeTypes::secp256k1_group::Fr,
+                                                       NativeTypes::secp256k1_group>(signing_message_str, account);
+    }
```
Will the typescript call this method (via a c_bind)? Or are we using some ethereum library to compute an ethereal signature in typescript?

suyash67 replied:
Added a comment here. So I am not using this function in TS for the following reason: the way TS computes the signature is:

message = hashTxRequest(txRequest)
signature = keyPair.signMessage(message)
So we don't really need to expose this function as a c_bind. I am only using this function to construct a signature for testing in C++.

iAmMichaelConnor commented on circuits/cpp/src/aztec3/circuits/kernel/private/.test.cpp:
```diff
@@ -284,12 +287,20 @@ PrivateInputs<NT> do_private_call_get_kernel_inputs(bool const is_constructor,
     //***************************************************************************
     Composer private_circuit_composer = Composer("../barretenberg/cpp/srs_db/ignition");
 
+    // TODO(Suyash): Oracle needs a bn254::fr private key so we are temporarily casting a secp256k1::fr
+    // private key to bn254::fr.
```
Sorry about this hacked-together oracle for testing purposes.

suyash67 replied:
No problem!

iAmMichaelConnor commented on circuits/cpp/src/aztec3/utils/types/circuit_types.hpp:
```diff
@@ -50,6 +51,10 @@ template <typename Composer> struct CircuitTypes {
 
+    typedef stdlib::secp256k1<Composer> secp256k1;
+    typedef typename secp256k1::bigfr_ct secp256k1_fr;
+    typedef typename secp256k1::g1_bigfr_ct secp256k1_point;
```
g1_bigfr_ct is a confusing name - I haven't looked at what this means:
Is it a group E(F_q), whose points are (Fq, Fq), which has group size Fr? (Where Fr is secp256k1_fr)
Or is it the group E(Fr), whose points are (Fr, Fr)?

suyash67 committed:
Compiles. Added eth public key and stuff.

suyash67 committed:
Fix account state tests.

suyash67 committed:
Use secp256k1 public key in synchroniser.

suyash67 committed:
Fix e2e zk token contract.

Post rebase fixes.

Update bb.

Rebase fixes.

suyash67 committed:
Fix new snapshots.

suyash67 committed:
Fix e2e zk test.

Co-authored-by: Álvaro Rodríguez <sirasistant@gmail.com>

suyash67 committed:
Use Grumpkin keystore while creating a simulation oracle.

suyash67 committed:
Fix clang tidy errors.

fix.

Fix clang tidy error.

suyash67 committed:
Move signing key to `SignedTxRequest`

TS changes.

`new` function in SignedTxRequest.

Co-authored-by: Álvaro Rodríguez <sirasistant@gmail.com>

test: fix unit tests with new async constructor

`signingKey` before `signature`.

Minor corrections.

suyash67 committed:
revert commented line in bootstrap.

suyash67 committed:
Move `usUtils` to foundation.

suyash67 committed:
Remove `EcdsaSignature` from circuits.js

suyash67 committed:
Remove unused imports.

suyash67 committed:
Use existing hash on `tx_request`.

suyash67 commented on circuits/cpp/src/aztec3/circuits/abis/signed_tx_request.hpp:
```diff
@@ -51,10 +46,40 @@ template <typename NCT> struct SignedTxRequest {
+        messages.push_back(tx_request.chain_id);
+
+        return NCT::compress(messages, aztec3::SIGN_TX_REQUEST);
+    }
```
Added a comment here. So I am not using this function in TS for the following reason: the way TS computes the signature is:
```TS
message = hashTxRequest(txRequest)
signature = keyPair.signMessage(message)
```
So we don't really need to expose this function as a c_bind. I am only using this function to construct a signature for testing in C++.

suyash67 replied:
Deleted this function as we can just hash `tx_request` to compute the message (instead of using a different generator).

suyash67 committed:
Fix keccak hashing in TS.

suyash67 committed:
Fix PKC.

suyash67 committed:
Correct ethereum address computation.

suyash67 committed:
Fix BigInt to string conversion for secp256k1 x-y.

suyash67 committed:
Update bb.

dbanks12 commented on circuits/cpp/src/aztec3/circuits/kernel/private/private_kernel_circuit.cpp:
```diff
@@ -261,6 +263,41 @@ void validate_inputs(PrivateInputs<CT> const& private_inputs)
     CT::fr const start_public_call_stack_length = array_length<Composer>(start.public_call_stack);
     CT::fr const start_new_l2_to_l1_msgs_length = array_length<Composer>(start.new_l2_to_l1_msgs);
 
+    {
+        // Verify public key hash matches ethereum address of the sender
+        // TODO(Suyash): We want to perform this only for the base case, so move to "initial" PKC.
+        const CT::address sender_address = private_inputs.signed_tx_request.tx_request.from;
+        const CT::secp256k1_point sender_public_key = private_inputs.signed_tx_request.signing_key;
+
+        // If the sender public key is P = (x, y), Ethereum address is computed as:
+        // E = keccak([x || y]).slice(12, 32)
+        CT::byte_array sender_public_key_bytes(sender_public_key.get_context());
+        sender_public_key_bytes.write(sender_public_key.x.to_byte_array());
+        sender_public_key_bytes.write(sender_public_key.y.to_byte_array());
+        const CT::byte_array sender_public_key_hash = stdlib::keccak<Composer>::hash(sender_public_key_bytes);
+        const CT::byte_array sender_address_bytes = CT::byte_array(sender_address.to_field());
+
+        // Check if the sender address matches the keccak hash of the public key
+        // Specifically, first 12 bytes must be 0, remaining 20 bytes must match.
+        for (size_t i = 0; i < 12; i++) {
+            sender_address_bytes[i].assert_is_zero(format("sender address at index ", i, " is non-zero"));
+        }
+        for (size_t i = 12; i < 32; i++) {
+            sender_address_bytes[i].assert_equal(
+                sender_public_key_hash[i], format("hash of public key does not match the sender address at index ", i));
+        }
+    }
+
+    // Verify signature against the first function being called (subsequent function calls do not
+    // need to be signed over by the sender, the sender only signs the inputs to the very first function)
+    const CT::byte_array message = private_inputs.signed_tx_request.tx_request.hash();
+    auto sig_verification_result = stdlib::ecdsa::verify_signature<Composer,
+                                                                   stdlib::secp256k1<Composer>,
+                                                                   stdlib::secp256k1<Composer>::fq_ct,
+                                                                   stdlib::secp256k1<Composer>::bigfr_ct,
+                                                                   stdlib::secp256k1<Composer>::g1_bigfr_ct>(
+        message, private_inputs.signed_tx_request.signing_key, private_inputs.signed_tx_request.signature);
+
```
Good work! Tbh our private kernel might be far behind enough that we might want to consider not maintaining it until we are ready to line-by-line compare with native and update it.

But for the ECDSA work, you needed to test the circuit-implementation so it is good to have this here :+1:

suyash67 replied:
Yeah exactly.

dbanks12 commented on circuits/cpp/src/aztec3/circuits/kernel/private/utils.cpp:
```diff
@@ -73,4 +76,19 @@ PreviousKernelData<NT> dummy_previous_kernel(bool real_vk_proof = false)
+    public_key_bytes.insert(public_key_bytes.end(), y_bytes.begin(), y_bytes.end());
+    std::vector<uint8_t> public_key_hash = stdlib::keccak<UltraComposer>::hash_native(public_key_bytes);
+    std::vector<uint8_t> chopped_public_key_hash(public_key_hash.size(), 0);
+    for (size_t i = 12; i < 32; i++) {
+        chopped_public_key_hash[i] = public_key_hash[i];
+    }
```
So the public key is all 0s except bytes 12 to 31 which are the public key hash? Can we link a doc here and/or add a comment?

suyash67 replied:
I've linked the reference and the section in it.

dbanks12 commented on yarn-project/key-store/src/secp256k1/test_key_store.ts:
```diff
@@ -0,0 +1,128 @@
+   *
+   * @returns A promise that resolves to an array of public keys associated with the accounts in the TestSecp256k1KeyStore.
+   */
+  public getSigningPublicKeys() {
+    return Promise.resolve(this.accounts.map(a => a.getPublicKey()));
+  }
```
Why `getSigningPublicKeys` vs `AccountPublicKey`?

suyash67 replied:
It's because `getSigningPublicKeys` was meant originally to get the signing key but we didn't really have signing keys until the ECDSA work. I don't want to change the key store interface so I am not changing this now, we can come back to rename this later (because the key store interface itself might change with the discussions around key management).

dbanks12 reviewed (approved):
Looks good to me! Made a few comments. Nothing major and I'm happy with a merge.

suyash67 commented on circuits/cpp/src/aztec3/circuits/kernel/private/native_private_kernel_circuit.cpp:
```diff
@@ -299,6 +301,45 @@ void validate_inputs(DummyComposer& composer, PrivateInputs<NT> const& private_i
+        const NT::secp256k1_point sender_public_key = private_inputs.signed_tx_request.signing_key;
+
+        // If the sender public key is P = (x, y), Ethereum address is computed as:
+        // E = keccak([x || y]).slice(12, 32)
```
Yes.

suyash67 commented on circuits/cpp/src/aztec3/circuits/kernel/private/native_private_kernel_circuit.cpp:
```diff
@@ -299,6 +301,45 @@ void validate_inputs(DummyComposer& composer, PrivateInputs<NT> const& private_i
     NT::fr const start_new_l2_to_l1_msgs_length = array_length(start.new_l2_to_l1_msgs);
 
+    {
+        // Verify public key hash matches ethereum address of the sender
+        // TODO(Suyash): We want to perform this only for the base case, so move to "initial" PKC.
```
Good point, it could be. In the current version, each of the private kernel is being passed on a `from` and `signing_key`, so I kept this check for each kernel iteration. But you're right, we could move it to just base case since that's what we will be doing eventually.

suyash67 replied:
Done.

suyash67 committed:
David's suggestions.

sirasistant commented on yarn-project/barretenberg.js/src/crypto/ecdsa/signature.ts:
```diff
@@ -0,0 +1,51 @@
+
+  /**
+   * Convert an ECDSA signature to a buffer.
+   * @returns A 65-character string of the form 0x<r><s><v>.
```
```suggestion
   * @returns A 130-character string of the form 0x<r><s><v>.
```

sirasistant commented on yarn-project/foundation/src/eth-public-key/index.ts:
```diff
@@ -0,0 +1,3 @@
+export * from './fr.js';
```
Could you re export this file from aztec-rpc/src/index.ts so we can import it in the e2e test through aztec.js instead of directly from foundation?

sirasistant reviewed (approved):
LGTM

iAmMichaelConnor commented:
I guess this PR has been superseded by 'account abstraction'?
Are there parts of this PR which would be useful to keep, for use by account contracts?

dbanks12 commented:
@suyash67 ^ thoughts?

dbanks12 commented:
No longer relevant in light of full account abstraction
